File: 01x10/c-005/main.php
<?php
function allReducable($num, $reducer) {
    for ($i = 1; $i <= $reducer; $i++) {
        if ($num % $i !== 0) {
            return FALSE;
        }
    }
    return TRUE;
}

function reduceResult($num, $reducer) {
    for ($i = $reducer; $i > 1; $i--) {
        if (allReducable($num / $i, $reducer)) {
            return reduceResult($num / $i, $reducer);
        }
    }
    return $num;
}

function smallestDivisible($reducer) {
    $result = 1;
    for ($i = 2; $i <= $reducer; $i++) {
        $result = reduceResult($result * $i, $i);
    }
    return $result;
}

print smallestDivisible(20);

File: 01x10/c-005/main.hs
allReducable num reducer = all (== 0) [mod num a| a <- [1..reducer]]

reduceResult num reducer
    | null result               = num
    | otherwise                 = reduceResult (head result) reducer
    where
        result = [div num a| a <- reverse [2..reducer], allReducable (div num a) reducer]

smallestDivisible reducer
    | reducer == 1              = 1
    | otherwise                 = reduceResult result reducer
    where
        result = smallestDivisible (reducer - 1) * reducer

main = do
    print $ smallestDivisible 20
